hasher.go

```diff
@@ -69,18 +74,76 @@ func (n *Hasher) NamespaceSize() namespace.IDSize {
 	return n.NamespaceLen
 }
 
-func NewNmtHasher(baseHasher hash.Hash, nidLen namespace.IDSize, ignoreMaxNamespace bool) *Hasher {
+func NewNmtHasher(baseHasher hash.Hash, nidLen namespace.IDSize, shareSize int, ignoreMaxNamespace bool) *Hasher {
 	return &Hasher{
-		Hash:             baseHasher,
-		NamespaceLen:     nidLen,
+		baseHasher:   baseHasher,
+		NamespaceLen: nidLen,
+		shareSize:    shareSize,
+
 		ignoreMaxNs:      ignoreMaxNamespace,
 		precomputedMaxNs: bytes.Repeat([]byte{0xFF}, int(nidLen)),
 	}
 }
 
+// Size returns the number of bytes Sum will return.
+func (n *Hasher) Size() int {
+	return n.baseHasher.Size() + int(n.NamespaceLen)*2
+}
+
+// Write writes the namespaced data to be hashed.
+//
+// Requires data of fixed size to match leaf or inner NMT nodes.
+// Only a single write is allowed.
+func (n *Hasher) Write(data []byte) (int, error) {
+	if n.data != nil {
+		panic("only a single Write is allowed")
+	}
+
+	ln := len(data)
+	switch ln {
+	default:
+		return 0, fmt.Errorf("invalid data size written to the hasher, len: %v", ln)
+	// inner nodes are made up of the nmt hashes of the left and right children
+	case n.Size() * 2:
+		n.tp = NodePrefix
+	// leaf nodes contain the namespace length and a share
+	case int(n.NamespaceLen) + n.shareSize:
```

Comment:
We might consider making leaf share a default case, then we wouldn't need to pass shareSize. 

It's ok to unlimit the share/leaf size as NMT can be used outside of the celestia'a use case, where leafs could have different sizes.

There is no security issues with this as integrity is checked anyway. As I can see so far no other panics would be introduced, but we should triple check that if decide leaf to be a default case.